### ISSUE TITLE
"Exit left" / "Exit right" instructions

### ISSUE THREAD

**Issue by devemux86** [abrensch/brouter#734]
Looking at the available turn instructions in [VoiceHint](https://github.com/abrensch/brouter/blob/master/brouter-core/src/main/java/btools/router/VoiceHint.java), we do not have "**Exit left**" and "**Exit right**" instructions.

They are very useful when we exit motorways. Now they are replaced by turn (slight) left / right instructions.

**Comment by afischerdev** [comment]
These would be very specific instructions, but they are indeed clearer
I made a first test. Just replace keep left/right with exit left/right on leaving motorways.
Something like this:
![exit_right](https://github.com/user-attachments/assets/95829066-3faf-4c05-bce7-7cd87c1e741f)

[testtrack0.2.gpx.txt](https://github.com/user-attachments/files/17083501/testtrack0.2.gpx.txt)

The problem will be the clients and their handling of the instructions.
In the test I changed all 'exit left/right' back to 'keep left/right' to be conform with the clients. 
But I guess that is not what you want.

**Comment by devemux86** [comment]
@afischerdev thanks for the example.

I do not understand what you mean with the change back?

GPX nav trk format has the advantage that is declares the instructions as strings.
Clients that already support exit instructions (like Cruiser) can use it immediately.

Nevertheless, "Exit left / right" are 2 new instructions.
So we will need new ids in `VoiceHint` and they must be included in all BRouter response types (like `geojson`).
Then clients must update their BRouter response parsing to handle the new instructions.

<img src="https://github.com/user-attachments/assets/104b5020-f632-4dda-8977-74374effa395" width="50%"/>

**Comment by afischerdev** [comment]
@devemux86 
> Nevertheless, "Exit left / right" are 2 new instructions.

Yes, that is what I did. And as you see it is working for the Locus export (timode=2). But for all other clients the new tag `ER / EL` have to go back to `KR / KL`.

**Comment by devemux86** [comment]
> But for all other clients

Do you mean other `turnInstructionMode`?

Also are exit instructions available in BRouter test website?

**Comment by devemux86** [comment]
@afischerdev 

> But for all other clients

Do you mean other `turnInstructionMode`?

Also are exit instructions available in BRouter test website?

**Comment by afischerdev** [comment]
@devemux86 
> Do you mean other turnInstructionMode?

Yes, and other exports.

> Also are exit instructions available in BRouter test website?

Sorry, no. Not at the moment, it was just a quick local test.

**Comment by devemux86** [comment]
@afischerdev 

> and other exports.

Will it be available in all output formats, gpx, geojson, etc.?

**Comment by afischerdev** [comment]
@devemux86 
> Will it be available in all output formats: gpx, geojson, etc.?

If we don't protect them for unexpected result, yes. Please give me some time, I'll make a more usable example.

**Comment by afischerdev** [comment]
Now a first version is ready. But before coming out, here a short discussion:

During testing an the simple rule (change from motorway to motorway_link) this [example](https://brouter.de/brouter-test/#map=12/53.3488/10.0599/standard&lonlats=10.07515,53.309597;10.046395,53.391561&profile=car-eco) comes up with `exit left`:

![exit](https://github.com/user-attachments/assets/a503086c-bccf-4bef-a4a3-3cef808b40c2)

[streetview first exit point](https://www.mapillary.com/app/?lat=53.36645762505&lng=10.033980473857014&z=17&pKey=398545767814419&focus=photo)
[streetview second exit point](https://www.mapillary.com/app/?lat=53.38894810775315&lng=10.039139783671999&z=17&pKey=296499628609288&focus=photo)

In my eyes the first point is not an exit, more a connection to next motorway.

And the BRouter database does not contain more info
point 1: highway=motorway_link surface=asphalt maxspeed=80 oneway=yes
point 2: highway=motorway_link maxspeed=100 oneway=yes

**Comment by devemux86** [comment]
@afischerdev

Do you mean that the exits should only be declared in instructions when change road categories?
i.e. change from motorway to motorway_link

**Comment by afischerdev** [comment]
@devemux86 
It should only be declared on motorway exits.
But the rule goes with `change from motorway to motorway_link` and so it catches what you see above.
So no idea how to grab that at the moment.

**Comment by devemux86** [comment]
@afischerdev 

Have you tested this area on other routers like [GraphHopper](https://graphhopper.com/maps), [OpenRouteService](https://maps.openrouteservice.org/), [Valhalla](https://valhalla.openstreetmap.de/)
to see what instructions they generate there?

**Comment by afischerdev** [comment]
@devemux86 
Both using keep left, keep right
Only Google brings an exit announcement.

**Comment by devemux86** [comment]
@afischerdev 

>  Both using keep left, keep right

Is it for the motorway_link to motorway_link example?

Because both routers have exit instructions.
Doing routings on motorway exits, they both produce exit instructions.

**Comment by afischerdev** [comment]
@devemux86 
> Is it for the motorway_link to motorway_link example?

Yes, it is. And Graphhopper as well uses keep left/right in this sample.

> Because both routers have exit instructions.
> Doing routings on motorway exits, they both produce exit instructions.

May be we find an example that contains exit in it. 
Think I found one:

[look here for BRouter](https://brouter.de/brouter-web/#map=15/47.4391/8.6960/standard&lonlats=8.638108,47.412465;8.69849,47.439772&profile=car-eco)
[look here for Valhalla - with exit](https://valhalla.openstreetmap.de/directions?profile=car&wps=8.637914657592775%2C47.41234909069777%2C8.69848966598511%2C47.43977169707289)
[look here for ORS - no exit, keep right](https://maps.openrouteservice.org/#/directions/Weidstrasse,Wangen%20b.%20D%C3%BCbendorf,ZH,Schweiz/Sportplatz,Effretikon,ZH,Schweiz/data/55,130,32,198,15,97,4,224,38,9,96,59,2,24,5,192,166,6,113,0,184,64,14,0,232,3,96,25,128,118,0,24,4,230,160,38,99,140,160,86,58,232,5,159,0,105,223,48,246,2,49,209,111,152,185,118,164,41,150,32,27,136,177,106,162,235,80,30,210,190,102,165,243,231,99,207,132,218,195,219,55,42,64,121,102,204,65,113,1,0,3,170,120,17,19,97,202,0,23,148,0,182,184,7,113,7,109,1,0,6,111,0,3,110,139,130,11,13,15,0,6,228,128,14,96,11,70,12,141,13,96,29,14,140,30,141,152,134,9,23,141,153,5,229,238,136,139,14,139,2,0,11,235,84,0,0)

However, for BRouter we have no additional data to map an exit. 
So it's either everything or nothing - everything is Exit or Keep left/right.

**Comment by devemux86** [comment]
@afischerdev 

> Graphhopper as well uses keep left/right in this sample.

GraphHopper does not support "exit" instructions (see [here](https://github.com/graphhopper/graphhopper/blob/master/web-api/src/main/java/com/graphhopper/util/Instruction.java)).

**Comment by devemux86** [comment]
@afischerdev 

Sorry about the OpenRouteService, it does not support "exit" instructions (see [here](https://github.com/GIScience/openrouteservice-docs#instruction-types)).

**Comment by afischerdev** [comment]
I think I found a solution for the 'exit problem' - the future voicehints are checked as well.
So here is a new output with exit Instruction only on last motorway point.
![exit_new](https://github.com/user-attachments/assets/7437b8a2-9c90-4844-9a68-562b626ff382)
